vlasovsolver/cpu_trans_map_amr.cpp

```diff
@@ -334,103 +334,111 @@ bool trans_map_1d_amr(const dccrg::Dccrg<spatial_cell::SpatialCell,dccrg::Cartes
+               int start = DimensionPencils[dimension].idsStart[pencili];
+               // Loop over cells in pencil
+               for (int b = 0; b < L; b++) {
+                  // Get cell pointer and local block id
+                  SpatialCell* srcCell = mpiGrid[DimensionPencils[dimension].ids[start + b]];
+                  const vmesh::LocalID blockLID = srcCell->get_velocity_block_local_id(blockGID,popID);
+                  // Store block data pointer for both loading of data and writing back to the cell
+                  if (blockLID != srcCell->invalid_local_id()) {
+                     // Get data pointer
+                     cellBlockData[start + b] = srcCell->get_data(blockLID,popID);
+                     nonEmptyBlocks++;
+                  } else {
+                     cellBlockData[start + b] = NULL;
+                  }
                }
+               if(nonEmptyBlocks == 0) {
+                  continue;
+               }
+               pencilBlocksCount.at(pencili) = nonEmptyBlocks;
+               // Transpose and copy block data from cells to source buffer
+               Vec* blockDataSource = blockDataBuffer.data() + start*WID3/VECL;
+               Realf** pencilBlockData = cellBlockData.data() + start;
+               copy_trans_block_data_amr(pencilBlockData, L, blockDataSource, vcell_transpose, popID);
             }
-            if(nonEmptyBlocks == 0) {
-               continue;
-            }
-            pencilBlocksCount.at(pencili) = nonEmptyBlocks;
-            // Transpose and copy block data from cells to source buffer
-            Vec* blockDataSource = blockDataBuffer.data() + start*WID3/VECL;
-            Realf** pencilBlockData = cellBlockData.data() + start;
-            copy_trans_block_data_amr(pencilBlockData, L, blockDataSource,
-                                      vcell_transpose, popID);
-         }
-         loadTimer.stop();
-
-         phiprof::Timer memsetTimer {memsetTimerId};
-         // reset blocks in all non-sysboundary neighbor spatial cells for this block id
-         for (CellID target_cell_id: DimensionTargetCells[dimension]) {
-            SpatialCell* target_cell = mpiGrid[target_cell_id];
-            if (target_cell) {
-               // Get local velocity block id
-               const vmesh::LocalID blockLID = target_cell->get_velocity_block_local_id(blockGID, popID);
-               // Check for invalid block id
-               if (blockLID != vmesh::VelocityMesh::invalidLocalID()) {
-                  // Get a pointer to the block data
-                  Realf* blockData = target_cell->get_data(blockLID, popID);
-                  memset(blockData, 0, WID3*sizeof(Realf));
+            loadTimer.stop();
+
+            phiprof::Timer memsetTimer {memsetTimerId};
+            // reset blocks in all non-sysboundary neighbor spatial cells for this block id
+            for (CellID target_cell_id: DimensionPencils[dimension].binsCells[activeBins[nBin]]) {
```

Comment:
There should be a per-bin collection of target cells to loop over here instead of cross-referencing the list of target cells and these sets. And that collection should be gathered once at the end of `binPencils`.

Reply:
Indeed, I hadn't noticed `DimensionTargetCells` is filled before binning. So in fact I can limit binning to cells that are potentially written into as well